### ISSUE TITLE
OpenID Connect support

### ISSUE THREAD

**Issue by guilhermednt** [bshaffer/oauth2-server-bundle#38]
It seems that `oauth2-server-php` supports OpenID Connect. Does it mean this bundle does it too?

**Comment by antonioberben** [comment]
+1

**Comment by bshaffer** [comment]
The bundle does support it, but you will access the OpenID Connect functions by accessing the `OAuth2\Server` object directly. The bundle does not support any specific OpenID Connect functionality itself.

**Comment by bshaffer** [comment]
I added some docs about how to set up [OpenID Connect with this library](http://bshaffer.github.io/oauth2-server-php-docs/overview/openid-connect/). I hope this helps.

**Comment by antonioberben** [comment]
Cheers buddy! I am playing with it already. Thanks a lot for all this. It is just great!

**Comment by bshaffer** [comment]
Thanks for the feedback. Not sure why it took me so long to add the docs. Please let me know where your pain points have been and I'll fill the docs out more to help with them.

**Comment by guilhermednt** [comment]
Thanks, @bshaffer!

I'm having a bit of trouble with the token endpoint (or maybe I didn't quite understand the specs).

I've managed to setup the bundle to use OIDC.

To test it I installed `oauth2-demo-php`. After testing a bit, I'm unsure if the behavior is correct on the token endpoint since I'm not getting the ID Token when I click the `Authorization Code` button at the OpenID Connect tab.

The spec says:

```
In addition to the response parameters specified by OAuth 2.0,
the following parameters MUST be included in the response:

    id_token
        ID Token value associated with the authenticated session.
```

But the response is:

``` json
{
  "access_token":"707495ba376dd3b6795ea6d06bb2f88314c9d2d7",
  "expires_in":3600,
  "token_type":"Bearer",
  "scope":"openid",
  "refresh_token":"2e750f52d70f6f19ec5952ca81e90e5d6012fd64"
}
```

To make the bundle work for OpenID Connect I created a Compiler Pass that adds `oauth2.storage.user_claims` and `oauth2.storage.public_key` services to the `oauth2.server` storage arguments. This Compiler Pass also changes the config to enable `use_openid_connect`.

Did I miss anything?

**Comment by bshaffer** [comment]
Hey there @guilhermednt!
Are you using `oauth2-demo-php` or `oauth2-server-bundle` in the case where it doesn't work? The bundle and the demo site are not compatible, as the bundle requires Symfony and the demo uses Silex. Could this be the root of your confusion?

**Comment by guilhermednt** [comment]
I have both:
1. a Symfony 2 app with the `oauth2-server-bundle` (it's my OpenID Provider) plus the stuff I did to make it accept OpenID Connect;
2. the `oauth2-demo-php` to interact with the Symfony app.

They are separate things...

**Comment by bshaffer** [comment]
So you're using the client side of the demo app, and the server side is
Symfony. Sounds good. But without seeing the code you've made for the
Symfony app, it's impossible for me to help. I would suggest making sure
the OpenID classes and controllers are properly created in your
Oauth2\Server class for starters.

**Comment by guilhermednt** [comment]
Thank you _very_ much for your attention, @bshaffer.

The code can be found [mostly here](https://github.com/PROCERGS/login-cidadao/tree/openid-connect/src/LoginCidadao/OpenIDBundle).

Basically I created the [compiler pass](https://github.com/PROCERGS/login-cidadao/blob/openid-connect/src/LoginCidadao/OpenIDBundle/DependencyInjection/Compiler/OverrideServiceCompilerPass.php#L22) to add the [storage classes](https://github.com/PROCERGS/login-cidadao/tree/openid-connect/src/LoginCidadao/OpenIDBundle/Storage) that I couldn't override via parameters, not sure if I did it right though...

Also, I had to [override the authorize endpoint to include the user id](https://github.com/PROCERGS/login-cidadao/blob/openid-connect/src/LoginCidadao/OpenIDBundle/Controller/AuthorizeController.php#L21).

Does this seem ok to you?
Again, thanks for the help!

**Comment by bshaffer** [comment]
Hmm... I unfortunately don't have time to debug this, but your configuration actually looks spot on. A few things I've found strange - none of my libraries seem to be in your `composer.json` file. Also, if you debug the `$server` object, can you confirm the compiler pass worked?

**Comment by guilhermednt** [comment]
The compiler pass seems to be working fine. I just debugged the `$server` to make sure. The lib is in the `composer.json` [here](https://github.com/PROCERGS/login-cidadao/blob/openid-connect/composer.json#L51).

I'm trying to find where the decision of sending an ID Token is being made but it's not very clear to me yet.

**Comment by bshaffer** [comment]
The decision is made by instantiating the [`OpenID` subclasses](https://github.com/bshaffer/oauth2-server-php/tree/develop/src/OAuth2/OpenID) of various `Controller`, `ResponseType`, `GrantType`, and `Storage` objects. These can be created and added manually to your `Server` object (using `setController`, `addGrantType`, etc), or they should be done automatically when `use_openid_connect` is set to true.

One thing that could cause an issue is if those various classes were being created automatically by the Server ahead of time. But it appears you set the container arguments before the server is instantiated, so I don't see how this would be the problem.

See if the [OpenID Authorization Code Grant Type](https://github.com/bshaffer/oauth2-server-php/blob/develop/src/OAuth2/OpenID/GrantType/AuthorizationCode.php#L24) ever gets executed, as this is where magic for the returning of the `id_token` happens in most cases.

**Comment by guilhermednt** [comment]
Hmm... This is not running. I'll take a look and try to find out why. If you have any suggestion it would be much appreciated.

Thanks for the help!

**Comment by bshaffer** [comment]
If I were you, I would debug in the Server class first and make sure the proper OpenID classes get created. 

https://github.com/bshaffer/oauth2-server-php/blob/develop/src/OAuth2/Server.php#L481
https://github.com/bshaffer/oauth2-server-php/blob/develop/src/OAuth2/Server.php#L490
https://github.com/bshaffer/oauth2-server-php/blob/develop/src/OAuth2/Server.php#L585
https://github.com/bshaffer/oauth2-server-php/blob/develop/src/OAuth2/Server.php#L594
https://github.com/bshaffer/oauth2-server-php/blob/develop/src/OAuth2/Server.php#L631

**Comment by guilhermednt** [comment]
Thanks! I'll take a look. One thing I noticed is that I had to override the `oauth2.grant_type.authorization_code.class` parameter to use `OAuth2\OpenID\GrantType\AuthorizationCode` but that wasn't enough since [this condition](https://github.com/bshaffer/oauth2-server-php/blob/develop/src/OAuth2/OpenID/GrantType/AuthorizationCode.php#L25) is failing.

Thanks a _LOT_ @bshaffer. I feel I'm in the correct path now.

**Comment by bshaffer** [comment]
Aha... yeah, we may be forcing some conditions through the bundle container injection that are making the OpenID model fail. I will do some testing as well to investigate this.

**Comment by guilhermednt** [comment]
Good news! I may have solved this last issue. The problem was in my `AuthorizationCode` storage class. The method `getAuthorizationCode()` wasn't including the `id_token` in the array.

**Comment by guilhermednt** [comment]
@bshaffer, another question: should the `nonce` be present [here](https://github.com/bshaffer/oauth2-server-bundle/blob/develop/Resources/views/Authorize/authorize.html.twig#L15)?

I noticed it's not being forwarded to the `id_token` since the `AuthorizeController` expects the `nonce` [in the query string](https://github.com/bshaffer/oauth2-server-php/blob/develop/src/OAuth2/OpenID/Controller/AuthorizeController.php#L58). I just want to be sure if this is the right place to put the `nonce`.

Again, thanks for the help! Your libs are saving my life. hehe

**Comment by bshaffer** [comment]
Ahh yes. OpenID Connect will still work for `response_type=code` without the nonce, but for `response_type=id_token` and `response_type=id_token+token`, we need the nonce to be passed in with the authorization request. I will submit a fix for this.

**Comment by bshaffer** [comment]
fixed in 7fba8128ade37a7d9c2e722c80f8fb9b26bb58fe
